lib/fabrik/VirtualHostManager.js

```diff
@@ -0,0 +1,168 @@
+const NotFound = errors.NotFound;
+const mapper = require('./VirtualHostRelationMapper');
+
+class VirtualHostManager extends BaseManager {
+
+  constructor(plan) {
+    super(plan);
+    this.director = bosh.director;
+    this.agent = new VirtualHostAgent(this.settings.agent);
+    this.mapper = mapper.VirtualHostRelationMapper;
+  }
+
+  verifyFeatureSupport(feature) {
+    if (!_.includes(this.agent.features, feature)) {
+      throw new NotImplemented(`Feature '${feature}' not supported`);
+    }
+  }
+
+  getDeploymentManifest(deploymentName) {
+    logger.info(`Fetching deployment manifest '${deploymentName}'...`);
+    return this.director
+      .getDeploymentManifest(deploymentName)
+      .tap(() => logger.info('+-> Fetched deployment manifest'))
+      .catch(err => {
+        logger.error('+-> Failed to fetch deployment manifest');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  getDeploymentIps(deploymentName) {
+    return this.getDeploymentManifest(deploymentName)
+      .tap(manifest => {
+        if (_.isNil(manifest)) {
+          throw new ServiceInstanceNotOperational(this.getInstanceGuid(deploymentName));
+        }
+      })
+      .then(manifest => _
+        .chain(manifest.jobs)
+        .map(job => _.map(job.networks, net => net.static_ips))
+        .flattenDeep()
+        .value()
+      );
+  }
+
+  createOrUpdateVirtualHost(deploymentName, params, guid) {
+    this.verifyFeatureSupport('multi_tenancy');
+    logger.info(`Creating virtual host for deployment '${deploymentName}'...`);
+    return this.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.provision(ips, guid))
+      .tap(() => this.saveVirtualHostInstanceToDeploymentRelation(deploymentName, guid))
+      .catch(err => {
+        logger.error('+-> Failed to create virtual host');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  deleteVirtualHost(deploymentName, guid) {
+    this.verifyFeatureSupport('multi_tenancy');
+    logger.info(`deleting virtual host for deployment '${deploymentName}'...`);
+    return this.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.deprovision(ips, guid))
+      .tap(() => this.deleteVirtualHostInstanceToDeploymentRelation(deploymentName, guid))
+      .catch(err => {
+        logger.error('+-> Failed to delete virtual host');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  createBinding(deploymentName, guid, binding) {
+    this.verifyFeatureSupport('credentials');
+    logger.info(`Creating binding '${binding.id}' for deployment '${deploymentName}', virtual host '${guid}'...`);
+    logger.info('+-> Binding parameters:', binding.parameters);
+    return this.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.createCredentials(ips, guid, binding.parameters))
+      .tap(credentials => this.createBindingProperty(deploymentName, binding.id, _.set(binding, 'credentials', credentials)))
```

Comment:
this should happen with promise chain with .then() - because if binding creds creation fails, then the operation should also be flagged as failed.